### PR TITLE
Extend tgw support prefix filters

### PR DESCRIPTION
**Pull request by adisos**
issue #330 
sub-task: initial support with unit-tests

### CODE REVIEW

**Review by zivnevo — COMMENTED** [review]
I think we need a json example from the resource-collector that contains prefix filters. Currently the new parsing code is not exercised in testing.

**Comment by adisos** [comment]
> I think we need a json example from the resource-collector that contains prefix filters, just to make sure we work correctly with real input.

Indeed, this is in a separate item on the issue description.

**Comment by zivnevo** [comment]
> > I think we need a json example from the resource-collector that contains prefix filters, just to make sure we work correctly with real input.
> 
> Indeed, this is in a separate item on the issue description.

ah, ok. I was too lazy to go to the issue :zany_face:

File: pkg/ibmvpc/parser.go [hunk 1]
```diff
@@ -610,16 +612,70 @@ func getTgwObjects(c *datamodel.ResourcesContainerModel,
...
+			fmt.Printf("warning: ignoring prefix filters, vpcID: %s, tgwID: %s, err is: %s\n", vpcUID, tgwUID, err.Error())
+		} else {
+			tgwMap[tgwUID].vpcFilters[vpc.ResourceUID] = permittedSubnets
+		}
 	}
 	return tgwMap
 }
 
+func isNodeNotFiltered(node vpcmodel.Node, filteredSubnets map[string]bool) bool {
+	// currently skipping nodes other than network interface (e.g. reserved ip, iks nodes)
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:633]: Why? Should there be a TODO here? A new task in the issue?

File: pkg/ibmvpc/parser.go [hunk 2]
```diff
@@ -610,16 +612,70 @@ func getTgwObjects(c *datamodel.ResourcesContainerModel,
...
+			}
+		case *Vsi:
+			if len(getNotFilteredNodes(n.nodes, filteredSubnets)) > 0 {
+				nodeSets = append(nodeSets, nodeSet)
+			}
+		case *VPC:
+			nodeSets = append(nodeSets, nodeSet)
+		}
+	}
+	return nodes, nodeSets, err
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:676]:
```suggestion
	return nodes, nodeSets, nil
```
More explicit...

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/tgw_analysis.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+func isSubnetMatchedByPrefixFilters(subnet *Subnet, tc *datamodel.TransitConnection) (bool, error) {
+	// Array of prefix route filters for a transit gateway connection. This is order dependent with those first in the
+	// array being applied first, and those at the end of the array is applied last, or just before the default.
+	pfList := tc.PrefixFilters
+
+	// Default setting of permit or deny which applies to any routes that don't match a specified filter.
+	pfDefault := tc.PrefixFiltersDefault
+
+	// TODO: currently assuming subnet cidrs are the advertised routes to the TGW, matched against the prefix filters
+	// TOTO: currently ignoring the "Before" field of each PrefixFilter, since assuming the array is ordered
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/tgw_analysis.go:40]: The documentation for the `PrefixFilters` field mentions the array is ordered:
```
// Array of prefix route filters for a transit gateway connection. This is order dependent with those first in the
// array being applied first, and those at the end of the array is applied last, or just before the default.
```

[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/ibmvpc/tgw_analysis.go:40]: OK. Makes sense. I guess we better validate that on the example we get from the cloud-resource-collector.

File: pkg/ibmvpc/tgw_analysis_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+						Prefix: &subnetD,
+					},
+					{
+						Action: &permit,
+						Prefix: &subnetE,
+					},
+					{
+						Action: &permit,
+						Prefix: &subnetE,
+						Le:     &le,
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/tgw_analysis_test.go:130]: added tests

**Review by zivnevo — APPROVED** [review]

**Pull request merged**